RecoMET/METAlgorithms/src/METSignificance.cc

```diff
@@ -51,54 +51,57 @@ metsig::METSignificance::getCovariance(const edm::View<reco::Jet>& jets,
    double cov_yy = 0;
  
    // for lepton and jet subtraction
-   std::vector<reco::CandidatePtr> footprint;
+   std::set<reco::CandidatePtr> footprint;
 
    // subtract leptons out of sumPt
    for ( std::vector< edm::Handle<reco::CandidateView> >::const_iterator lep_i = leptons.begin();
          lep_i != leptons.end(); ++lep_i ) {
-      for( reco::CandidateView::const_iterator lep = (*lep_i)->begin(); lep != (*lep_i)->end(); lep++ ){
-         if( lep->pt() > 10 ){
-            for( unsigned int n=0; n < lep->numberOfSourceCandidatePtrs(); n++ ){
-               if( lep->sourceCandidatePtr(n).isNonnull() and lep->sourceCandidatePtr(n).isAvailable() ){
-                  footprint.push_back(lep->sourceCandidatePtr(n));
-               } 
-            }
-         }
-      }
+     for( reco::CandidateView::const_iterator lep = (*lep_i)->begin(); lep != (*lep_i)->end(); lep++ ){
+       if( lep->pt() > 10 ){
+	 for( unsigned int n=0; n < lep->numberOfSourceCandidatePtrs(); n++ ){
+	   if( lep->sourceCandidatePtr(n).isNonnull() and lep->sourceCandidatePtr(n).isAvailable() ){
+	     footprint.insert(lep->sourceCandidatePtr(n));
+	   } 
+	 }
+       }
+     }
    }
    // subtract jets out of sumPt
    for(edm::View<reco::Jet>::const_iterator jet = jets.begin(); jet != jets.end(); ++jet) {
 
-      // disambiguate jets and leptons
-      if(!cleanJet(*jet, leptons) ) continue;
+     // disambiguate jets and leptons
+     if(!cleanJet(*jet, leptons) ) continue;
+     for( unsigned int n=0; n < jet->numberOfSourceCandidatePtrs(); n++){
+       if( jet->sourceCandidatePtr(n).isNonnull() and jet->sourceCandidatePtr(n).isAvailable() ){
 
-      for( unsigned int n=0; n < jet->numberOfSourceCandidatePtrs(); n++){
-         if( jet->sourceCandidatePtr(n).isNonnull() and jet->sourceCandidatePtr(n).isAvailable() ){
-            footprint.push_back(jet->sourceCandidatePtr(n));
-         }
-      }
+	 footprint.insert(jet->sourceCandidatePtr(n));
+       }
+     }
 
    }
 
    // calculate sumPt
    double sumPt = 0;
-   for( edm::View<reco::Candidate>::const_iterator cand = pfCandidates.begin();
-         cand != pfCandidates.end(); ++cand){
-
-      // check if candidate exists in a lepton or jet
-      bool cleancand = true;
-      for(unsigned int i=0; i < footprint.size(); i++){
-         if( footprint[i]->p4() == cand->p4() ){
-            cleancand = false;
-         }
-      }
-      // if not, add to sumPt
-      if( cleancand ){
-         sumPt += cand->pt();
-      }
-
+   for(size_t i = 0; i< pfCandidates->size();  ++i) {
+     
+     // check if candidate exists in a lepton or jet
+     bool cleancand = true;
+     if(footprint.find( pfCandidates->ptrAt(i) )==footprint.end()) {
+
+       //dP4 recovery
+       for( std::set<reco::CandidatePtr>::const_iterator it=footprint.begin();it!=footprint.end();it++) {
+	 if( ((*it)->p4()-(*pfCandidates)[i].p4()).Rho()<0.005 ){
```

Comment:
There is first a matching by reference, then a matching per pt. The motivation is that if for the miniAOD everything looks OK, for the computation on the AOD objects, the matching between leptons and pf Candidates seems to not be the same, thus leading to different results between a MET significance computed on AOD and miniAOD.
We mitigated the effect with that p4/pT recovery.
Going to use Pt2 instead of current p4